src/util/psbt/mod.rs

```diff
@@ -208,6 +217,471 @@ impl PartiallySignedTransaction {
     }
+
+    /// Attempts to create _all_ the required signatures for this PSBT using `k`.
+    ///
+    /// If you just want to sign an input with one specific key consider using `sighash_ecdsa`. This
+    /// function does not support scripts that contain `OP_CODESEPARATOR`.
+    ///
+    /// # Panics
+    /// If an attempt is made to sign a taproot input.
+    pub fn sign<C: Signing>(&mut self, k: impl GetKey + Copy, secp: &Secp256k1<C>) -> Result<(), SignError> {
+        let tx = self.unsigned_tx.clone();
+        let mut cache = SighashCache::new(&tx);
+
+        for i in 0..self.inputs.len() {
+            match self.signing_algorithm(i).expect("i is valid") {
+                SigningAlgorithm::Ecdsa => self.bip32_sign_ecdsa(k, i, &mut cache, secp)?,
+                SigningAlgorithm::Schnorr => unimplemented!("taproot PSBT signing not yet implemented"), 
+            }
+        }
+        Ok(())
+    }
+
+    /// Attempts to create all signatures required by this PSBT's `bip32_derivation` field, adding
+    /// them to `partial_sigs`.
+    ///
+    /// ## Returns
+    /// If an error is returned some signatures may already have been added to the PSBT. Since
+    /// `partial_sigs` is a [`BTreeMap`] it is safe to retry, previous sigs will be overwritten.
+    fn bip32_sign_ecdsa<C: Signing, T: Deref<Target=Transaction>>(
+        &mut self,
+        k: impl GetKey,
+        input_index: usize,
+        cache: &mut SighashCache<T>,
+        secp: &Secp256k1<C>
+    ) -> Result<(), SignError> {
+        let (msg, sighash_ty) = self.sighash_ecdsa(input_index, cache)?;
+        let input = &self.inputs[input_index]; // index checked in call to `sighash_ecdsa`;
+        let mut sigs = Vec::with_capacity(input.bip32_derivation.len());
+
+        for (pk, key_source) in input.bip32_derivation.iter() {
+            let sk = if let Ok(Some(sk)) = k.get_key(KeyRequest::Bip32(key_source.clone()), Some(secp)) {
+                sk
+            } else if let Ok(Some(sk)) = k.get_key(KeyRequest::Pubkey(PublicKey::new(*pk)), Some(secp)) {
+                sk
+            } else {
+                return Err(SignError::KeyNotFound);
+            };
+
+            let sig = EcdsaSig {
+                sig: secp.sign_ecdsa(&msg, &sk.inner),
+                hash_ty: sighash_ty,
+            };
+
+            let pk = sk.public_key(secp);
+            sigs.push((pk, sig));
+        }
+
+        for (pk, sig) in sigs.iter() {
+            self.inputs[input_index]
+                .partial_sigs
+                .insert(*pk, *sig);
+        }
+
+        Ok(())
+    }
+
+    /// Returns the sighash message to sign an ECDSA input along with the sighash type.
+    ///
+    /// Uses the [`EcdsaSighashType`] from this input if one is specified. If no sighash type is
+    /// specified uses [`EcdsaSighashType::All`]. This function does not support scripts that
+    /// contain `OP_CODESEPARATOR`.
+    ///
+    /// # Panics
+    /// If called with a taproot input.
```

Comment:
I would let others chime in on this too. I don't like having panic on functions that already return a result.

Reply:
Good point, I added a new error variant `SignError::WrongSigningAlgorithm`.

Reply:
The doc comment about panic is still here.